onnxruntime/core/optimizer/insert_cast_transformer.cc

```diff
@@ -231,37 +231,29 @@ static Status ForceSingleNodeCPUFloat16ToFloat32(onnxruntime::Graph& graph, cons
   return Status::OK();
 }
 
-enum TypeGroup {
-  Unknown = -1,
-  Bool = 0,
-  Integer = 1,
-  Float = 2,
-};
-
-TypeGroup GetTypeGroup(DataType type) {
-  if (*type == "tensor(bool)") {
-    return Bool;
-  }
-
-  if (*type == "tensor(int16)" || *type == "tensor(int32)" || *type == "tensor(int64)" || *type == "tensor(int8)" ||
-      *type == "tensor(uint16)" || *type == "tensor(uint32)" || *type == "tensor(uint64)" || *type == "tensor(uint8)") {
-    return Integer;
-  }
-
-  if (*type == "tensor(bfloat16)" || *type == "tensor(double)" || *type == "tensor(float)" || *type == "tensor(float16)") {
-    return Float;
-  }
-
-  return Unknown;
-}
-
 /** Transformer to remove duplicate Cast nodes. */
 class RemoveDuplicateCastTransformer : public GraphTransformer {
  public:
   RemoveDuplicateCastTransformer() : GraphTransformer("RemoveDuplicateCastTransformer") {
   }
 
  private:
+  static constexpr std::array<std::string_view, 13> cast_ordering{
+      "tensor(bool)", "tensor(uint8)", "tensor(uint16)", "tensor(uint32)", "tensor(uint64)", "tensor(int8)", "tensor(int16)",
```

Comment:
why the order is `uint8, uint16, uint32, uint64, int8, int16, int32, int64`. 
For example, uint64 to int8 conversion will loss precision (might overflow).

Reply:
Good point, I don't think there is a topological ordering here. Will revert.

Reply:
> instead of uint8, int8, uint16, int16, uint32, int32, uint64, int64?

I suppose this has the issue that uin8 -> int8 is not safe.
I don't think there is a topological ordering here. Will revert to the bit length check I had before.